module/applications/activity/activity-sheet.mjs

```diff
@@ -0,0 +1,494 @@
+   * @this {ActivityConfig}
+   * @param {Event} event         Triggering click event.
+   * @param {HTMLElement} target  Button that was clicked.
+   */
+  static #addRecovery(event, target) {
+    const periods = new Set(
+      Object.entries(CONFIG.DND5E.limitedUsePeriods).filter(([, config]) => !config.deprecated).map(([k]) => k)
+    );
+    const existingPeriods = new Set(this.activity.uses.recovery.map(t => t.period));
+    const filteredPeriods = periods.difference(existingPeriods);
+    this.activity.update({
+      "uses.recovery": [
+        ...this.activity.toObject().uses.recovery,
+        { period: filteredPeriods.first() ?? periods.first() }
+      ]
+    });
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Handle removing an entry from the consumption targets list.
+   * @this {ActivityConfig}
+   * @param {Event} event         Triggering click event.
+   * @param {HTMLElement} target  Button that was clicked.
+   */
+  static #deleteConsumption(event, target) {
+    const consumption = this.activity.toObject().consumption.targets;
+    consumption.splice(target.closest("[data-index]").dataset.index, 1);
+    this.activity.update({ "consumption.targets": consumption });
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Handle removing an entry from the uses recovery list.
+   * @this {ActivityConfig}
+   * @param {Event} event         Triggering click event.
+   * @param {HTMLElement} target  Button that was clicked.
+   */
+  static #deleteRecovery(event, target) {
+    const recovery = this.activity.toObject().uses.recovery;
+    recovery.splice(target.closest("[data-index]").dataset.index, 1);
+    this.activity.update({ "uses.recovery": recovery });
+  }
+
+  /* -------------------------------------------- */
+  /*  Form Handling                               */
+  /* -------------------------------------------- */
+
+  /**
+   * Handle form submission.
+   * @param {SubmitEvent} event          Triggering submit event.
+   * @param {HTMLFormElement} form       The form that was submitted.
+   * @param {FormDataExtended} formData  Data from the submitted form.
+   */
+  static async #onSubmitForm(event, form, formData) {
+    const submitData = this._prepareSubmitData(event, formData);
+    await this._processSubmitData(event, submitData);
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Perform any pre-processing of the form data to prepare it for updating.
+   * @param {SubmitEvent} event          Triggering submit event.
+   * @param {FormDataExtended} formData  Data from the submitted form.
+   * @returns {object}
+   */
+  _prepareSubmitData(event, formData) {
+    const submitData = foundry.utils.expandObject(formData.object);
+    if ( foundry.utils.hasProperty(submitData, "consumption.targets") ) {
+      submitData.consumption.targets = Object.values(submitData.consumption.targets);
```

Comment:
I think this might be handled automatically by the DataModel now, see `ArrayField#_cast`. But I'm not sure, so this is fine to leave as well.

Reply:
It only seemed to be partially working for me, I would be able to update the `<select>` inputs perfectly find, but as soon as I entered something into the associated text `<input>` the select value would get reset. I couldn't quite figure out what was happening but explicitly casting the data to arrays before saving fixed it.